subprojects/core/src/test/groovy/org/gradle/api/internal/AbstractDomainObjectCollectionSpec.groovy

```diff
@@ -1431,4 +1420,339 @@ abstract class AbstractDomainObjectCollectionSpec<T> extends Specification {
+        }
+    }
+
+    abstract static class AbstractConfigurationClosureFactory extends AbstractConfigurationFactory<Closure> {
+        static String getConfigurationType() {
+            return Closure.simpleName
+        }
+
+        @Override
+        Closure create() {
+            return { AbstractConfigurationClosureFactory.this.callNoReentrant() }
+        }
+    }
+
+    class AddActionFactory extends AbstractConfigurationActionFactory {
+        @Override
+        void call() {
+            container.add(b)
+        }
+    }
+
+    class AddClosureFactory extends AbstractConfigurationClosureFactory {
+        @Override
+        void call() {
+            container.add(b)
+        }
+    }
+
+    class AddLaterActionFactory extends AbstractConfigurationActionFactory {
+        AddLaterActionFactory() {
+            containerAllowsExternalProviders()
+        }
+
+        @Override
+        void call() {
+            container.addLater(Providers.of(b))
+        }
+    }
+
+    class AddLaterClosureFactory extends AbstractConfigurationClosureFactory {
+        AddLaterClosureFactory() {
+            containerAllowsExternalProviders()
+        }
+
+        @Override
+        void call() {
+            container.addLater(Providers.of(b))
+        }
+    }
+
+    class AddAllLaterActionFactory<T> extends AbstractConfigurationActionFactory<T> {
+        AddAllLaterActionFactory() {
+            containerAllowsExternalProviders()
+        }
+
+        @Override
+        void call() {
+            container.addAllLater(new CollectionProviderInternal<T, Collection<T>>() {
+                @Override
+                Class getElementType() {
+                    return type
+                }
+
+                @Override
+                int size() {
+                    return 1
+                }
+
+                @Override
+                Class getType() {
+                    return List
+                }
+
+                @Override
+                ProviderInternal<T> map(Transformer transformer) {
+                    throw new UnsupportedOperationException()
+                }
+
+                @Override
+                Collection<T> get() {
+                    return [b]
+                }
+
+                @Override
+                Collection<T> getOrNull() {
+                    return get()
+                }
+
+                @Override
+                Collection<T> getOrElse(Collection<T> defaultValue) {
+                    return get()
+                }
+
+                @Override
+                boolean isPresent() {
+                    return true
+                }
+            })
+        }
+    }
+
+    class AddAllLaterClosureFactory<T> extends AbstractConfigurationClosureFactory {
+        AddAllLaterClosureFactory() {
+            containerAllowsExternalProviders()
+        }
+
+        @Override
+        void call() {
+            container.addAllLater(new CollectionProviderInternal<T, Collection<T>>() {
```

Comment:
Is this anonymous type different than the anonymous type in the previous class?  If not, then let's pull it out and reuse the definition - then if there are changes to the interfaces, it doesn't require multiple mechanical updates.